hibernate-core/src/test/java/org/hibernate/cache/spi/ReadWriteCacheTest.java

```diff
@@ -0,0 +1,310 @@
+			log.info( "Update Book using HQL" );
+			int numRows = session.createQuery( "update Book set title = :title where id = :id" )
+					.setParameter( "title", UPDATED_TITLE )
+					.setParameter( "id", bookId )
+					.executeUpdate();
+			assertEquals( 1, numRows );
+			interceptTransaction.set( true );
+		} );
+
+		endLatch.await();
+		interceptTransaction.set( false );
+
+		doInHibernate( this::sessionFactory, session -> {
+			loadBook( bookId, session );
+		} );
+	}
+
+	@Test
+	@TestForIssue(jiraKey = "HHH-13792")
+	public void testUpdateNativeQuery() throws InterruptedException {
+		bookId = 6L;
+
+		doInHibernate( this::sessionFactory, session -> {
+			createBook( bookId, session );
+		} );
+
+		doInHibernate( this::sessionFactory, session -> {
+			log.info( "Update Book using NativeQuery" );
+			int numRows = session.createNativeQuery( "update Book set title = :title where id = :id" )
+					.setParameter( "title", UPDATED_TITLE )
+					.setParameter( "id", bookId )
+					.addSynchronizedEntityClass( Book.class )
+					.executeUpdate();
+			assertEquals( 1, numRows );
+			interceptTransaction.set( true );
+		} );
+
+		endLatch.await();
+		interceptTransaction.set( false );
+
+		doInHibernate( this::sessionFactory, session -> {
+			loadBook( bookId, session );
+		} );
+	}
+
+	@Override
+	protected Class<?>[] getAnnotatedClasses() {
+		return new Class<?>[] {
+				Book.class,
+		};
+	}
+
+	@Override
+	protected String getCacheConcurrencyStrategy() {
+		return "read-write";
+	}
+
+	private void assertBookNotFound(long bookId, Session session) {
+		log.info( "Load Book" );
+		Book book = session.get( Book.class, bookId );
+		assertNull( book );
+	}
+
+	private void createBook(long bookId, Session session) {
+		log.info( "Create Book" );
+		Book book = new Book();
+		book.setId( bookId );
+		book.setTitle( ORIGINAL_TITLE );
+		session.save( book );
+	}
+
+	private Consumer<Configuration> getCacheConfig() {
+		return configuration -> configuration.setInterceptor( new TransactionInterceptor() );
+	}
+
+	private void loadBook(long bookId, Session session) {
+		log.info( "Load Book" );
+		Book book = session.get( Book.class, bookId );
+		assertNotNull( book );
+		assertEquals( "Found old value", UPDATED_TITLE, book.getTitle() );
+	}
+
+	@Entity(name = "Book")
+	@Cacheable
+	@org.hibernate.annotations.Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
+	private static final class Book {
+
+		@Id
+		private Long id;
+
+		private String title;
+
+		public Long getId() {
+			return id;
+		}
+
+		public void setId(Long id) {
+			this.id = id;
+		}
+
+		public String getTitle() {
+			return title;
+		}
+
+		public void setTitle(String title) {
+			this.title = title;
+		}
+
+		public String toString() {
+			return "Book[id=" + id + ",title=" + title + "]";
+		}
+	}
+
+	private final class TransactionInterceptor extends EmptyInterceptor {
+		@Override
+		public void beforeTransactionCompletion(Transaction tx) {
+			if ( interceptTransaction.get() ) {
```

Comment:
If this value is ``true`` then a separate session will load the entity here:
- (_without the fix_) the cache was already cleared by this point, so the old value gets added to the cache and will remain after the transaction is committed
- (_with the fix_) the cache is cleared after the transaction is committed - so even though the old value is loaded here it will be evicted after this method completes